pydantic_ai_slim/pydantic_ai/_agent_graph.py

```diff
@@ -1039,14 +1054,38 @@ async def _call_tools(
         projected_usage.tool_calls += len(tool_calls)
         usage_limits.check_before_tool_call(projected_usage)
 
+    # Checks for soft limits(if any set on total tools)
+    can_make_tool_calls = tool_manager.can_make_tool_calls(len(tool_calls), deepcopy(usage))
+
+    calls_to_run: list[_messages.ToolCallPart] = []
+
+    # For each tool, check how many calls are going to be made
+    tool_call_counts: defaultdict[str, int] = defaultdict(int)
+    _projection_count_of_tool_usage(tool_call_counts, tool_calls)
+
     for call in tool_calls:
-        yield _messages.FunctionToolCallEvent(call)
+        current_tool_use = tool_manager.get_current_use_of_tool(call.tool_name)
```

Comment:
Could this be managed entirely in `ToolManager`? That can already short-circuit what is returned for a given tool call

Reply:
Refactored into tool_manager.

> It would be interesting to implement this suggestion from https://github.com/pydantic/pydantic-ai/issues/3352#issuecomment-3507572317 at the same time:

I don't think we should just use the previous call's output though, it gives incorrect implication to the model that the response was from that set of args(if any) and I am also not sure about allowing some of the parallel calls and denying the rest. It does move away from UsageLimits behaviour that people might be familiar with. 
We could make this configurable(all-nothing / calls within limit allowed others returned) and default to all-nothing which is same as UsageLimits?